#!/bin/bash
#SBATCH -J sqanti_indel
#SBATCH -D /projects/dveiga/analysis/git/BRCA_isoforms/sqanti
#SBATCH -N 4
#SBATCH -n=16
#SBATCH --mem 128G
#SBATCH -t 0-72:00:00
#SBATCH -p compute
#SBATCH -q batch
#SBATCH -o %x.out
#SBATCH -e %x.out

#source /projects/banchereau-lab/ISO-seq/annotation_processing/pitchfork_ToFU2_dev/setup-env.sh
dir_blast=/projects/banchereau-lab/tools/blast/2_6_0/bin
dir_hmmer=/projects/banchereau-lab/tools/hmmer3/3_1/bin
dir_gmap=/projects/banchereau-lab/tools/gmap_gsnap/2015_12_31/bin
dir_r=/projects/banchereau-lab/tools/3_4_4/bin
PATH="$dir_blast:$dir_hmmer:$dir_gmap:$dir_r:$PATH"

# ***** Requirements ****

## Input Data files
#These files are generated in PacBio

ISOFORMS_GTF="/projects/dveiga/analysis/git/BRCA_isoforms/PacBio_Breast_cancer_allTranscripts.gtf"
REF_GENOME="/projects/zhaoy/Human_Mouse_Genome/Human/hg38.fa" #uses chr1, chr2,
REF_GTF="/projects/banchereau-lab/tools/Gencode_30/gencode.v30.annotation.gtf"
# Gmap index of Reference
GMAP_INDEX="/projects/banchereau-lab/ISO-seq/annotation_processing/GRCh38/release-88/gmap-gsnap-2015-12-31-chr-indices/hg38"

# **** RUN STEPS *****

# Running SQANTI gmap alignment (indel correction)
python /projects/banchereau-lab/tools/sqanti/sqanti_qc.py -g -t 96 -x $GMAP_INDEX $ISOFORMS_GTF $REF_GTF $REF_GENOME
